### ISSUE TITLE
Clean up Numba PRNG

### ISSUE THREAD

**Issue by agayev** [numba/numba#2782]
NumPy and Python both have Mersenne Twister PRNG implementation, but they differ in (1) how they initialize the state and (2) generate random numbers.  Numba, while trying to support both implementations imitates the initialization from NumPy and random number generation from Python.  As a result, neither is easily supported.  

For example, the following code seeds the PRNG with value of 2 and generates a pair of random numbers, both in NumPy and Numba-compiled NumPy.  Ideally, both implementations should generate the same sequence of random numbers and the assertion at the bottom should succeed.  Also, for the Python counterpart of the same code, where we raplace np.random.seed() with random.seed(), and np.random.randint() with random.randrange(), the assertion should succeed.

```Python
from numba import jit
import numpy as np

seed = 2
max = 100

np.random.seed(seed)
a = [np.random.randint(max), np.random.randint(max)]

@jit(nopython=True)
def foo():
    np.random.seed(seed)
    return [np.random.randint(max), np.random.randint(max)]

b = foo()

assert a == b
```

However, currently, in both versions the assertion fails.  Since Numba imitiates initialization from NumPy and random number generation from Python, there are ugly hacks in numba/numba/test_random.py that first generate a Python PRNG state and then manually copies the state to Numba's internal state, after which Numba can generate the same sequence of random numbers as Python.

The following patch, changes the Numba's random number generation to match that of NumPy's.  As a result, with the following patch, the assertion in the above code succeds:

```Python
diff --git a/numba/targets/randomimpl.py b/numba/targets/randomimpl.py
index 150541e..99cd61d 100644
--- a/numba/targets/randomimpl.py
+++ b/numba/targets/randomimpl.py
@@ -148,7 +148,9 @@ def get_next_int(context, builder, state_ptr, nbits):
     def get_shifted_int(nbits):
         shift = builder.sub(c32, nbits)
         y = get_next_int32(context, builder, state_ptr)
-        return builder.lshr(y, builder.zext(shift, y.type))
+        mask = builder.not_(ir.Constant(y.type, 0))
+        mask = builder.lshr(mask, builder.zext(shift, y.type))
+        return builder.and_(y, mask)
```

Hence, with the following patch, Numba's PRNG follows NumPy's PRNG verbatim with one caveat: this does not handle the case of iflarge branch in get_next_int -- more research is needed to figure out how to make that branch also match NumPy's PRNG, although it should be trivial.

In conclusion, the cleanup work needs to have another PRNG implementation that keeps current number generation but modifies the initialization to match that of Python's.  With this work in place, Numba would have two different PRNG implementations that are identical to Python's and NumPy's.  Then the unit tests should be fixed to remove the ugly state-setting hacks and have straightforward test implementations similar to the above code snippets.

On another thought, given that most Numba users are compiling NumPy code, perhaps Numba can drop support for Python PRNG and just keep NumPy PRNG.

**Comment by stuartarchibald** [comment]
@agayev thanks very much for debugging and writing this up. Adding a bit to this... The debugging discussion, which may be useful later, is here: https://gitter.im/numba/numba-dev?at=5a96d65d8f1c77ef3a6816a5
Useful info gleaned from this conversation...
 * Numba src https://github.com/numba/numba/blob/master/numba/targets/randomimpl.py#L151
 * Numpy src https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/randomkit.c#L303
 * Both use `rk_random` as per the original Mersenne Twister description. From the result, Numpy takes the last N bits, Numba takes the first N bits. Watch out for tempering and where that is applied.
 * As noted above, the branching for larger than 32bit type PRNG isn't handled in the patch above, NumPy src needs closer inspection to make sure the shift is applied in Numba is appropriate to match.
 * It should be possible to support both NumPy and Python PRNGs from the same code base, branching to chose N bits from MSB/LSB end.
 * As noted above, the unittests will need a rewrite and during this it may be a good idea to improve state manipulation tooling.
